File: graph/backTrack/basic/combinationSumII.py
# find unique combination
# Each number in candidates may only be used once in the combination.
'''
Input: candidates = [2,5,2,1,2], target = 5,  12225
A solution set is:
[
  [1,2,2],
  [5]
]
'''
def combinationSum2(self, candidates, target):
    res = []
    self.dfs(sorted(candidates), res, [], target, 0)
    return res

def dfs(self, candidates, res, path, target, index):
    if target == 0:
        res.append(path)
        return
    if target < 0:
        return
    for i in range(index, len(candidates)):
        if i > index and candidates[i] == candidates[i-1]:  # no duplicate at the same level   124a)  124b)
            continue
        self.dfs(candidates, res, path + [candidates[i]], target - candidates[i], i+1)  # i+1 makes it not using the smae canidates again


def combinationSum2(self, candidates, target):
    res = []
    self.dfs(sorted(candidates), res, [], target, 0)
    return res

def dfs(self, candidates, res, path, target, index):
    if target == 0:
        res.append(path)
        return
    if target < 0:
        return
    visited = set()
    for i in range(index, len(candidates)):
        if candidates[i] not in visited:
            visited.add(candidates[i])
            self.dfs(candidates, res, path + [candidates[i]], target - candidates[i], i+1)

File: graph/backTrack/basic/combinationSumII.java
//combination sumII
public List<List<Integer>> combinationSum2(int[] candidates, int target) {
    List<List<Integer>> res = new ArrayList<List<>>();
    Arrays.sort(candidates);
    dfs(res, new ArrayList<Integer>(), candidates, target, 0);
    return res;
}

private void dfs(List<List<Integer>> res,
                       List<Integer> path,
                       int[] candidates,
                       int remain,
                       int start){
    if(remain < 0)
        return;
    else if(remain == 0)
        res.add(new ArrayList<>(path));
    else{
        for(int i = start; i < candidates.length; i++){
            //Each number in C may only be used once in the combination.
            if(i == start || candidates[i] != candidates[i-1]) {
              path.add(candidates[i]);
              dfs(res, path, candidates, remain - candidates[i], i + 1);
              path.remove(path.size() - 1);
            }
        }
    }
}
